utils/binutils.ml

```diff
@@ -0,0 +1,577 @@
+      number_of_sections: int;
+      size_of_optional_header: int;
+      characteristics: int;
+    }
+
+  let read_header e_lfanew d buf =
+    let number_of_sections = get_uint16 d buf 6 in
+    let size_of_optional_header = get_uint16 d buf 20 in
+    let characteristics = get_uint16 d buf 22 in
+    {e_lfanew; number_of_sections; size_of_optional_header; characteristics}
+
+  type optional_header_magic =
+    | PE32
+    | PE32PLUS
+
+  type optional_header =
+    {
+      magic: optional_header_magic;
+      image_base: int64;
+    }
+
+  let read_optional_header d {e_lfanew; size_of_optional_header; _} =
+    let buf = load_bytes d Int64.(add e_lfanew (of_int header_size)) size_of_optional_header in
+    let magic =
+      match get_uint16 d buf 0 with
+      | 0x10b -> PE32
+      | 0x20b -> PE32PLUS
+      | n -> raise (Error (Unsupported ("optional header magic", n)))
+    in
+    let image_base =
+      match magic with
+      | PE32 -> unsigned_of_int32 (get_uint32 d buf 28)
+      | PE32PLUS -> get_uint64 d buf 24
+    in
+    {magic; image_base}
+
+  type section =
+    {
+      name: string;
+      virtual_size: int;
+      virtual_address: int64;
+      size_of_raw_data: int;
+      pointer_to_raw_data: int64;
+    }
+
+  let section_header_size = 40
+
+  let read_sections d {e_lfanew; number_of_sections; size_of_optional_header; _} =
+    let buf =
+      load_bytes d
+        Int64.(add e_lfanew (of_int (header_size + size_of_optional_header)))
+        (number_of_sections * section_header_size)
+    in
+    let mk i =
+      let base = i * section_header_size in
+      let name = name_at ~max_len:8 buf (base + 0) in
+      let virtual_size = get_uint d buf (base + 8) in
+      let virtual_address = unsigned_of_int32 (get_uint32 d buf (base + 12)) in
+      let size_of_raw_data = get_uint d buf (base + 16) in
+      let pointer_to_raw_data = unsigned_of_int32 (get_uint32 d buf (base + 20)) in
+      {name; virtual_size; virtual_address; size_of_raw_data; pointer_to_raw_data}
+    in
+    Array.init number_of_sections mk
+
+  type symbol =
+    {
+      name: string;
+      address: int64;
+    }
+
+  let load_section_body d {size_of_raw_data; pointer_to_raw_data; _} =
+    load_bytes d pointer_to_raw_data size_of_raw_data
+
+  let find_section sections sectname =
+    array_find (function ({name; _} : section) -> name = sectname) sections
+
+  let read_symbols d {image_base; _} sections =
+    match find_section sections ".exptbl" with
+    | None -> [| |]
+    | Some ({virtual_address; _} as exptbl) ->
+        let buf = load_section_body d exptbl in
+        let numexports = Int64.to_int (get_word d buf 0) in
+        let word_size = word_size d in
+        let mk i =
+          let address = get_word d buf (word_size * (2 * i + 1)) in
+          let nameoff = get_word d buf (word_size * (2 * i + 2)) in
+          let name =
+            let off = Int64.(sub nameoff (add virtual_address image_base)) in
```

Comment:
I guess this a virtual address (VA) to relative virtual address (RVA) translation, but I don't really know in which address space  `nameoff` and `address` are specified. Would maybe a few comments be in order ?

Reply:
I think so too, but it took me a couple of tries to arrive at the right formula. Like the following two points, this encoding is private to `FlexDLL`. I will see if I find something cogent to write about it.